src/gcp_airflow_foundations/source_class/ftp_gcs_source.py

```diff
@@ -0,0 +1,217 @@
+            objects=files_to_wait_for,
+            task_group=taskgroup
+        )
+
+    def load_to_landing_task(self, table_config, taskgroup):
+        return PythonOperator(
+            task_id="load_gcs_to_landing_zone",
+            op_kwargs={"table_config": table_config},
+            python_callable=self.load_to_landing_py_op_task,
+            task_group=taskgroup
+        )
+
+    def load_to_landing_py_op_task(self, table_config, **kwargs):
+        data_source = self.config.source
+
+        ti = kwargs['ti']
+        ds = kwargs['ds']
+
+        files_to_load = ti.xcom_pull(key='file_list', task_ids='ftp_taskgroup.get_file_list')
+        logging.info(files_to_load)
+
+        # Parameters
+        bucket = self.config.source.extra_options["gcs_bucket"]
+        source_format = table_config.extra_options.get("gcs_table_config")["source_format"]
+        gcp_project = data_source.gcp_project
+        landing_dataset = data_source.landing_zone_options.landing_zone_dataset 
+        destination_table = f"{gcp_project}:{landing_dataset}.{table_config.landing_zone_table_name_override}" + f"_{ds}"
+
+        schema_file_name = None
+        if "schema_file" in table_config.extra_options.get("gcs_table_config"):
+            schema_file_name = table_config.extra_options.get("gcs_table_config")["schema_file"]
+
+        # Get files to load from metadata file
+        if schema_file_name:
+            schema_file = self.gcs_hook.download(bucket_name=bucket, object_name=schema_file_name)
+
+            # Only supports json schema file format - add additional support if required
+            schema_fields = json.loads(schema_file)
+            logging.info(schema_fields)
+            gcs_to_bq = GCSToBigQueryOperator(
+                task_id='import_files_to_bq_landing',
+                bucket=bucket,
+                source_objects=files_to_load,
+                source_format=source_format,
+                schema_fields=schema_fields,
+                destination_project_dataset_table=destination_table,
+                write_disposition='WRITE_TRUNCATE',
+                create_disposition='CREATE_IF_NEEDED',
+                skip_leading_rows=1,
+            )
+        else:
+            gcs_to_bq = GCSToBigQueryOperator(
+                task_id='import_files_to_bq_landing',
+                bucket=bucket,
+                source_objects=files_to_load,
+                source_format=source_format,
+                destination_project_dataset_table=destination_table,
+                write_disposition='WRITE_TRUNCATE',
+                create_disposition='CREATE_IF_NEEDED',
+                skip_leading_rows=1,
+            )
+        gcs_to_bq.execute(context=kwargs)
+
+    def get_list_of_files(self, table_config, **kwargs):
+        # XCom push the list of files
+        bucket = self.config.source.extra_options["gcs_bucket"]
+        if "metadata_file" in table_config.extra_options.get("gcs_table_config"):
```

Comment:
Just thinking out loud. 
Metadata or a single templated file are not the only options.  Some other options I can think of are (a) list all files in a folder (after the success/metadata file is detected), (b) list all files with some prefi 

Not need to implement all of them right now. But thinking what's the best way to structure the config to make it extendable 

Maybe having a "file_discovery_type" enum config, and then doing something different for each (can be done in the base class). 

Based on which enum is selected, we can validate the rest of the options: if MetadataFile metadata_file must not be null, etc.

Reply:
Yeah, that would make sense - can add it to the to-dos

Reply:
we technically already have prefix support - would just need to add folder support and some additional logic to make it all work